include/cadmium/engine/pdevs_dynamic_coordinator.hpp

```diff
@@ -0,0 +1,247 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef CADMIUM_PDEVS_DYNAMIC_COORDINATOR_HPP
+#define CADMIUM_PDEVS_DYNAMIC_COORDINATOR_HPP
+
+#include <cadmium/modeling/message_bag.hpp>
+#include <cadmium/modeling/dynamic_coupled.hpp>
+#include <cadmium/engine/pdevs_dynamic_simulator.hpp>
+#include <cadmium/engine/pdevs_dynamic_engine.hpp>
+#include <cadmium/modeling/dynamic_message_bag.hpp>
+#include "pdevs_dynamic_engine_helpers.hpp"
+
+namespace cadmium {
+    namespace dynamic {
+        namespace engine {
+
+            template<typename TIME, typename LOGGER>
+            class coordinator : public cadmium::dynamic::engine::engine<TIME> {
+                //TODO(Lao): implement correct coordinator formatter that does not depend on the MODEL template parameter
+                //using formatter=typename cadmium::logger::coordinator_formatter<MODEL, TIME>;
+
+                //MODEL is assumed valid, the whole model tree is checked at "runner level" to fail fast
+                TIME _last; //last transition time
+                TIME _next; // next transition scheduled
+
+                std::string _model_id;
+
+                subcoordinators_type<TIME> _subcoordinators;
+                external_couplings<TIME> _external_output_couplings;
+                external_couplings<TIME> _external_input_couplings;
+                internal_couplings<TIME> _internal_coupligns;
+
+            public:
+
+                dynamic::message_bags _inbox;
+                dynamic::message_bags _outbox;
+
+                using model_type = typename cadmium::dynamic::modeling::coupled<TIME>;
+
+                /**
+                 * @brief A dynamic coordinator must be constructed with the coupled model to coordinate.
+                 */
+                coordinator() = delete;
+
+                coordinator(std::shared_ptr<model_type> coupled_model) :
+                        _model_id(coupled_model->get_id())
+                {
+
+                    std::map<std::string, std::shared_ptr<engine<TIME>>> models_by_id;
+
+                    for(auto& m : coupled_model->_models) {
+                        std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> m_coupled = std::dynamic_pointer_cast<cadmium::dynamic::modeling::coupled<TIME>>(m);
+                        std::shared_ptr<cadmium::dynamic::modeling::atomic_abstract<TIME>> m_atomic = std::dynamic_pointer_cast<cadmium::dynamic::modeling::atomic_abstract<TIME>>(m);
+
+                        if (m_coupled == nullptr) {
+                            assert(m_atomic != nullptr);
```

Comment:
This would me that someone defined a model that is child of coupled and atomic at the same time. This should break completely.

Reply:
Yes, that is true, but i don't see the problem:
1) If someone defines a model a child of atomic and coupled, it can't be a well defined model and the assert(m_atomic != nullptr) catches that case and simulation is over as it should.
2) In a normal model programing flow for cadmium::dynamic, an atomic model is made using the class cadmium::dynamic::modeling::atomic<MODEL, TIME> as a wrapper class, and this class will throw an static_assert using cadmium::concept::is_atomic<ATOMIC>::value, in other hand the coupled cadmium::dynamic::coupled class has other validations. thus, the program flow would already failed when creating the shared_ptr before reaching this point.

Reply:
I added a verification that the pointer did not correctly cast to both atomic and coupled to do not let users to do that.